File: 03- Subprogramas y recursividad/recursividad/python/tabla_multiplicar.py
""" 
Descripción: Realiza un programa que solicite al usuario un número entero y cree un
    método recursivo que regrese su tabla de multiplicar del 1 al 10.
Autor: Santiago Nicolás De la mora Núñez
Fecha de creación: 11/04/2026
"""

def main():
    # Solicita un número
    numero = solicitar_numero()

    print("Impresión recursiva")
    imprimir_tabla_multiplicar_recursivamente(numero, 1)

    print("\nImpresión iterativa")
    imprimir_tabla_multiplicar_iterativamente(numero)

# Solicita el ingreso de un número entero positivo
def solicitar_numero():
    numero = ""

    while True:
        numero = input("Ingrese un número positivo: ")

        # Verifica si no se ingresó un númeor o se dejó el campo vacío
        if numero.strip() == "":
            print("\nError. No se puede dejar el campo vacío.\n")
            continue

        # Captura errores
        try:
            # Convierte la cadena en un número entero
            numero = int(numero)

            # Verifica si el número es positivo
            if numero > 0:
                return numero
            else:
                print("\nError. No puede ingresar números negativos ni 0.\n")
        except ValueError:
            print("\nError. No puede ingresar letras ni carácteres especiales.\n")

def imprimir_tabla_multiplicar_recursivamente(n, i):
    # Caso recursivo
    if i <= 10:
        print(f"{n} x {i} = {n * i}")
        imprimir_tabla_multiplicar_recursivamente(n, i + 1)
    
def imprimir_tabla_multiplicar_iterativamente(n):
    i = 1
    while i <= 10:
        print(f"{n} x {i} = {n * i}")

        i += 1

# Ejecuta el programa
if __name__ == '__main__':
    main()

File: 03- Subprogramas y recursividad/recursividad/java/TablaMultiplicar.java
/**
 * Descripción: Realiza un programa que solicite al usuario un número entero y cree un
 *      método recursivo que regrese su tabla de multiplicar del 1 al 10.
 * Autor: Santiago Nicolás De la mora Núñez
 * Fecha de creación: 16/03/2026
 */

import javax.swing.JOptionPane;

public class TablaMultiplicar {
    public static void main(String[] args) {
        // Declaración de variables
        int numero;

        // Solicita un número
        numero = solicitarNumero();

        // Verifica si el usuario decidió cancelar la solicitud del número
        if (numero != 0) {
            System.out.println("Impresión recursiva");
            imprimirTablaMultiplicarRecursivamente(numero, 1);

            System.out.println("\nImpresión iterativa");
            imprimirTablaMultiplicarIterativamente(numero);
        }
    }

    // Solicita el ingreso de un número entero positivo
    public static int solicitarNumero() {
        String numeroStr;
        int numero;

        while (true) {
            numeroStr = JOptionPane.showInputDialog(null, "Ingrese un número positivo", "Número positivo", JOptionPane.PLAIN_MESSAGE);

            // Verifica si se hizo click en el tache o en "Cancelar" (si el valor es null)
            if (numeroStr == null) {
                return 0;
            }

            // Verifica si se no ingresó un número o se dejó el campo vacío
            if (numeroStr.trim().isEmpty()) {
                JOptionPane.showMessageDialog(null, "No puede dejar el campo vacío.", "Error", JOptionPane.ERROR_MESSAGE);
                continue;
            }

            // Captura errores
            try {
                // Convierte la cadena en un número entero
                numero = Integer.parseInt(numeroStr);

                // Verifica si el número es positivo
                if (numero > 0) {
                    return numero;
                } else {
                    JOptionPane.showMessageDialog(null, "No puede ingresar números negativos ni 0.", "Error", JOptionPane.ERROR_MESSAGE);
                }
            } catch (NumberFormatException e) {
                JOptionPane.showMessageDialog(null, "No puede ingresar letras ni carácteres especiales.", "Error", JOptionPane.ERROR_MESSAGE);
            }
        }
    }

    public static void imprimirTablaMultiplicarRecursivamente(int n, int i) {
        // Caso recursivo
        if (i <= 10) {
            System.out.println(n + " x " + i + " = " + (n * i));
            imprimirTablaMultiplicarRecursivamente(n, i + 1);
        }
    }

    public static void imprimirTablaMultiplicarIterativamente(int n) {
        int i = 1;
        while (i <= 10) {
            System.out.println(n + " x " + i + " = " + (n * i));
            
            i++;
        }
    }
}